api/turing/api/router_versions_api.go

```diff
@@ -33,6 +34,46 @@ func (c RouterVersionsController) ListRouterVersions(
 	return Ok(routerVersions)
 }
 
+// CreateRouterVersion creates a router version from the provided configuration. If no router exists
+// within the provided project with the provided id, this method will throw an error.
+// If the update is valid, a new RouterVersion will be created but NOT deployed.
+func (c RouterVersionsController) CreateRouterVersion(r *http.Request, vars RequestVars, body interface{}) *Response {
```

Comment:
I expect that more of existing code can be re-used here. Current implementation of the `Edit Router` feature, CREATES a new router version and DEPLOYS it immediately. This new `CreateRouterVersion` does the same except that it doesn't deploy the new version after the new version is saved into the DB. Which makes me thinking that that for both usecases, the same controller method can be used, with extra `?deploy=true|false` parameter passed to distinguish between these two scenarios.

Reply:
I have a similar but somewhat different view here. At the moment, the corresponding services for these entities is mostly a data access layer. I feel more of the business logic should be moved there and the controller layer kept lean to only handle parsing the request payload, calling the service methods and returning the final response in the appropriate format (i.e., all the presentation details).

So, update router API will call:
1. routerService.CreateRouter
2. routerVersionService.CreateRouterVersion
3. deploymentService.DeployVersion // handles updating router states, current version of router entity

Create router version API will only need to call service methods 1 & 2.

This still allows reusability but keeping the controllers very intentional. On the downside, this is going to take some refactoring. We did this exercise for experiments after the first cut (when considering the possibility of versioning the APIs) and I think the code structure is cleaner.  If we agree to take this route for Turing as well, we can add it to the backlog (and leave this controller method as is for now).

Reply:
@romanwozniak @deadlycoconuts thoughts?

Reply:
Hmm that sounds possible but like you said it's going to require quite a big amount of refactoring to reduce the controller layer to something of a request handler layer. I don't really have an opinion on an option that I prefer because I don't really have the context that the ones who decided on the relative responsibilities the controller/service layers should have, so I'm open to anything.

Reply:
The proposal makes total sense for me, but I don't think that it should be in the scope of this PR as it requires major refactoring of the other controller/service parts. So [this](https://github.com/gojek/turing/pull/179#discussion_r826757089) change is still preferred by me for now

Reply:
Alright. I don't think the refactoring should be in this PR either, will add it to the backlog.